File: core/java/src/main/java/hara/lang/protocol/IIndexedKV.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.iindexedkv", name = "IIndexedKV")
public interface IIndexedKV<K, V> {
  @HaraMethod(value = "index-of-key", arity = 2)
  long indexOfKey(K key);

  @HaraMethod(value = "index-of-val", arity = 2)
  long indexOfVal(V val);
}

File: core/rust/src/lang/protocol/iindexedkv.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.iindexedkv", name = "IIndexedKV")]
pub trait IIndexedKV<K, V> {
    #[hara_method(value = "index-of-key", arity = 2)]
    fn index_of_key(&self, key: &K) -> Option<usize>;
    #[hara_method(value = "index-of-val", arity = 2)]
    fn index_of_val(&self, value: &V) -> Option<usize>;
}
